### PR TITLE
2292 control channel client api client

### LINKED ISSUE #2292: Add IslandAPIClient to the agent
# Refactor

## Component(s) to be refactored

- control.py
- control_channel.py
- network/relay/utils.py
- caching_agent_binary_repository.py
- monkey.py
- agent_event_forwarder.py

## Explanation

Lots of different components are communicating with the Island's HTTP API. They're all issuing HTTP requests using `requests`. Ideally, these components wouldn't be focused on such low-level details.

Add `IslandAPIClient` that wraps all of the API calls that are relevant to the agent. There should be a 1-to-1 mapping between the methods of `IslandAPIClient` and the Island's HTTP API. Refactor all other components that are communicating with the Island to use the `IslandAPIClient`

### Notes
1. The constructor of `IslandAPIClient` should accept a server and attempt to connect. If connection fails, it should raise some kind of `ConnectionError`. This will simplify `find_server()`.
2. `control_channel.py` will likely be just a thin wrapper around `IslandAPIClient`. That's fine, since ISP states that "clients should not be forced to depend upon interfaces that they do not use", then `IControlChannel` serves as a view to segregate `AutomatedMaster` from parts of `IslandAPIClient` it doesn't use. In time we can reevaluate this.
3. `control.py` will likely end up as another thin wrapper. That's fine, it's only still there to provide backwards compatibility with old components that still need it.
4. `register_agent()` doesn't belong in `ControlChannel`. Whatever is calling it now could just call `IslandAPIClient` instead.
5. Serialization/deserialization should be the responsibility of `IslandAPIClient`. All other components can think in terms of objects.

## Tasks
- [x] Define a set of Errors/Exceptions that `IslandAPIClient` can raise. (0d) @ilija-lazoroski @VakarisZ 
- [x] Define the `IslandApIClient` (0d) @VakarisZ @shreyamalviya 
    - [x] Define a constructor that takes a server string and raises an exception if it cannot connect
- [x] Modify `find_server()` logic to use the new `IslandAPIClient` (0d) @ilija-lazoroski 
- [x] Refactor control.py - @shreyamalviya 
     - [x] Implement `send_agent_log()` and modify ControlClient.send_log() to use `IslandApiClient` (0d)
     - [x] Implement `get_pba_file()` and modify ControlClient.get_pba_file() to use `IslandApiClient` (0d)
- [ ] Refactor control_channel.py @cakekoa 
    - [x] Implement `register_agent()` and remove it from `ControlChannel`. Use `IslandAPIClient.register_agent()` in monkey.py instead of `control_channel.register_agent()` (0.25d)
    - [x] Implement `should_agent_stop()` and modify `ControlChannel` to use `IslandAPIClient` (0.25d)
    - [x] Implement `get_config()` and modify `ControlChannel` to use `IslandAPIClient` (0.25d)
    - [x] Implement `get_credentials_for_propagation()` and modify `ControlChannel` to use `IslandAPIClient` (0.25d)
- [x] Implement `get_agent_binary()` and modify `CachingAgentBinaryRepository` to use `IslandAPIClient` (0d) - @shreyamalviya 
- [x] Implement `send_events()` and modify `AgentEventForwarder` to use `IslandAPIClient` (0d) @ilija-lazoroski

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Fixes #2292.

Moves functionality from `ControlClient` to `HTTPIslandAPIClient`.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] ~~Was the CHANGELOG.md updated to reflect the changes?~~
* [ ] ~~Was the documentation framework updated to reflect the changes?~~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [ ] Have you successfully tested your changes locally? Elaborate:
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [ ] ~~If applicable, add screenshots or log transcripts of the feature working~~

### CODE REVIEW

File: monkey/infection_monkey/island_api_client/http_island_api_client.py [hunk 1]
```diff
@@ -27,7 +35,15 @@ def handle_island_errors(fn):
...
-        except requests.exceptions.ConnectionError as err:
+        except (
+            IslandAPIConnectionError,
+            IslandAPIError,
+            IslandAPIRequestError,
+            IslandAPIRequestFailedError,
+            IslandAPITimeoutError,
+        ) as e:
+            raise e
+        except (requests.exceptions.ConnectionError, requests.exceptions.TooManyRedirects) as err:
```
Comment (mssalvatore) [inline monkey/infection_monkey/island_api_client/http_island_api_client.py:46]: These all inherit from `IslandAPIError`, so you can just 

```suggestion
        except IslandAPIError as err:
            raise err
        except (requests.exceptions.ConnectionError, requests.exceptions.TooManyRedirects) as err:
```

Also, use `err` not `e` for consistency. It's written somewhere in our guidelines.

File: monkey/infection_monkey/island_api_client/http_island_api_client.py [hunk 2]
```diff
@@ -116,6 +143,59 @@ def send_events(self, events: Sequence[JSONSerializable]):
...
+        url = f"{self._api_url}/monkey-control/needs-to-stop/{agent_id}"
+        response = requests.get(  # noqa: DUO123
+            url,
+            verify=False,
+            timeout=SHORT_REQUEST_TIMEOUT,
+        )
+        response.raise_for_status()
+
+        json_response = json.loads(response.content.decode())
+        return json_response["stop_agent"]
```
Comment (mssalvatore) [inline monkey/infection_monkey/island_api_client/http_island_api_client.py:169]:
```suggestion
        return response.json()["stop_agent"]
```

File: monkey/infection_monkey/island_api_client/http_island_api_client.py [hunk 3]
```diff
@@ -116,6 +143,59 @@ def send_events(self, events: Sequence[JSONSerializable]):
...
+    @convert_json_error_to_island_api_error
+    def get_config(self) -> AgentConfiguration:
+        response = requests.get(  # noqa: DUO123
+            f"{self._api_url}/agent-configuration",
+            verify=False,
+            timeout=SHORT_REQUEST_TIMEOUT,
+        )
+        response.raise_for_status()
+
+        config_dict = json.loads(response.text)
```
Comment (mssalvatore) [inline monkey/infection_monkey/island_api_client/http_island_api_client.py:181]:
```suggestion
        config_dict = response.json()
```

File: monkey/infection_monkey/master/control_channel.py [hunk 4]
```diff
@@ -21,11 +25,30 @@
...
+
 class ControlChannel(IControlChannel):
-    def __init__(self, server: str, agent_id: str):
+    def __init__(self, server: str, agent_id: str, api_client: IIslandAPIClient):
         self._agent_id = agent_id
         self._control_channel_server = server
+        self._island_api_client = api_client
 
+    @handle_island_api_errors
     def register_agent(self, parent: Optional[UUID] = None):
```
Comment (mssalvatore) [inline monkey/infection_monkey/master/control_channel.py:52]: This should get removed from `ControlClient`

File: monkey/infection_monkey/master/control_channel.py [hunk 5]
```diff
@@ -21,11 +25,30 @@
...
+    def wrapper(*args, **kwargs):
+        try:
+            print(args)
+            func(*args, **kwargs)
+        except (
+            IslandAPIConnectionError,
+            IslandAPIRequestError,
+            IslandAPIRequestFailedError,
+            IslandAPITimeoutError,
+        ) as e:
```
Comment (mssalvatore) [inline monkey/infection_monkey/master/control_channel.py:39]: See above comment about simplifying this.

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/island_api_client/http_island_api_client.py (hunk 1 above)
Comment (cakekoa) [reply monkey/infection_monkey/island_api_client/http_island_api_client.py:46]: Nice!

**Pull request merged**